File: moduleLike/src/main/java/com/example/modulelike/core/ModuleProvider.java
package com.example.modulelike.core;

import com.example.modulelike.utils.ReflectUtils;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by chenyy on 2022/2/16.
 */

public class ModuleProvider {

    private static HashMap<String, BaseAppLike> moduleLikes = new HashMap<>();

    public static void register(Class clazz) {
        moduleLikes.put(clazz.getName(), loadModuleLike(clazz));
    }

    public static void init() {
        try {
            Class clazz = Class.forName(Const.MODULE_LOADER_INIT);
            ReflectUtils.invokeStaticMethod(clazz, Const.INIT_METHOD);
            for (Map.Entry<String, BaseAppLike> entry : moduleLikes.entrySet()) {
                if (entry != null && entry.getValue() != null) entry.getValue().onCreate();
            }
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
    }

    private static BaseAppLike loadModuleLike(Class clazz) {
        try {
            BaseAppLike appLike = (BaseAppLike) clazz.newInstance();
            return appLike;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: Lib/moduleLike/src/main/java/com/example/modulelike/core/ModuleProvider.kt
package com.example.modulelike.core

import com.example.modulelike.utils.ReflectUtils

/**
 * Created by chenyy on 2022/2/18.
 */

object ModuleProvider {
    private val moduleLikes = mutableMapOf<String, BaseAppLike?>()

    fun register(clazz: Class<*>) {
        moduleLikes[clazz.name] = loadModuleLike(clazz)
    }

    fun init() {
        try {
            val clazz = Class.forName(Const.MODULE_LOADER_INIT)
            ReflectUtils.invokeStaticMethod<Any>(clazz, Const.INIT_METHOD)
            moduleLikes.forEach {
                it.value?.onCreate()
            }
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }

    private fun loadModuleLike(clazz: Class<*>): BaseAppLike? {
        try {
            return clazz.newInstance() as BaseAppLike
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }
}